magpie/api/login/login.py

```diff
@@ -271,69 +279,127 @@ def login_success_external(request, external_user_name, external_id, email, prov
                          http_kwargs={"location": homepage_route, "headers": headers})
 
 
+def user_from_token(request, token):
+    # type: (Request, Str) -> Optional[models.User]
+    """
+    Return the ``User`` associated with the token as long as the token is valid and issued by this Magpie instance.
+    """
+    magpie_secret = get_constant("MAGPIE_SECRET", request, settings_name="magpie.secret")
+    decoded_token = jwt.decode(token, magpie_secret, algorithms="HS256", issuer=MAGPIE_INSTANCE_NAME,
+                               options={"require": ["exp", "token"]})
+
+    network_token = (request.db.query(models.NetworkToken)
+                     .join(models.User)
+                     .filter(models.NetworkToken.token == decoded_token["token"])
+                     .first())
+    if network_token:
+        return network_token.user
+
+
 @s.ProviderSigninAPI.get(schema=s.ProviderSignin_GET_RequestSchema, tags=[s.SessionTag],
                          response_schemas=s.ProviderSignin_GET_responses)
 @view_config(route_name=s.ProviderSigninAPI.name, permission=NO_PERMISSION_REQUIRED)
-def authomatic_login_view(request):
```

Comment:
Yeah that's a great idea. I looked into it a bit but this is not my area of expertise at all. If you have some suggestions for where to start, what libraries to use, etc. then I can start looking into this.

An initial change with JWTs was much more feasible for me

Reply:
The main hurdle for me was how to set up authomatic providers that can be changed dynamically (ie. the same as creating a new `Network Node`)

Reply:
I think directly using `authomatic` by itself could work. It would take care of performing the OAuth requests flow with the relevant remote Magpie Network Node reference. 

The only thing that would be needed is to implement the endpoints to support OAuth 2.0, such as `/authorize`, `/userinfo`, etc. That is better explained in the spec https://datatracker.ietf.org/doc/html/rfc6749. 

I've not tried `authomatic` with JWT yet, but how to use them is well-defined in the standard (https://oauth.net/2/jwt-access-tokens/), so I would imagine it can support it natively. If not, it's maybe just a matter of adding the class that implements it, similar to the `esgfopenid` and `wso2` implementations in https://github.com/Ouranosinc/Magpie/blob/master/magpie/security.py and https://github.com/Ouranosinc/Magpie/tree/master/magpie/api/login

To make it dynamic with Magpie Network Nodes, I think expanding `authomatic_setup` with a set of configs passed as input would be sufficient. 

I've not played with the `authomatic` code for a while now. I would have to remind myself as well of all its features 😅

Reply:
Ok this is super helpful. I'll start exploring this possibility this week and see how far I can get.